include/stumpless/error.h

```diff
@@ -30,85 +30,100 @@
 extern "C" {
 #  endif
 
+
 /**
- * An identifier of the types of errors that might be encountered.
- *
- * Note that the same error may be encountered in different contexts with a
- * different ultimate cause. To completely diagnose a problem, you will need
- * to look at more than just the error id.
+ * A macro to implement the enum values && enum names string array.
  */
-enum stumpless_error_id {
-/**
+#define FOREACH_ERROR(ERROR)	\
+/** \var STUMPLESS_ADDRESS_FAILURE
  * A provided network address was not valid.
  *
  * This might be formatting problem, or it might mean that a DNS lookup failed.
- */
-  STUMPLESS_ADDRESS_FAILURE,
-/** A required function parameter was NULL or empty. */
-  STUMPLESS_ARGUMENT_EMPTY,
-/** A provided argument was too big, for example to fit into a buffer target. */
-  STUMPLESS_ARGUMENT_TOO_BIG,
+ */\
+  ERROR(STUMPLESS_ADDRESS_FAILURE, 0)	\
```

Comment:
I was able to get documentation generated using your approach, but you will need to update the Doxyfile at `tools/doxygen/Doxyfile.in` in order to make it work. I have included the change below - it consists of moving two settings into the 'deliberate settings' section and making sure that they are both set to `YES`.

```
diff --git a/tools/doxygen/Doxyfile.in b/tools/doxygen/Doxyfile.in
index a4b6389..e209077 100644
--- a/tools/doxygen/Doxyfile.in
+++ b/tools/doxygen/Doxyfile.in
@@ -13,6 +13,8 @@ USE_MDFILE_AS_MAINPAGE = @PROJECT_SOURCE_DIR@/README.md
 OPTIMIZE_OUTPUT_FOR_C  = YES
 SEARCH_INCLUDES        = YES
 INCLUDE_PATH           = @PROJECT_SOURCE_DIR@/include @PROJECT_BINARY_DIR@/include
+ENABLE_PREPROCESSING   = YES
+MACRO_EXPANSION        = YES
 
 # default generated settings
 BRIEF_MEMBER_DESC      = YES
@@ -154,8 +156,6 @@ GENERATE_AUTOGEN_DEF   = NO
 GENERATE_PERLMOD       = NO
 PERLMOD_LATEX          = NO
 PERLMOD_PRETTY         = YES
-ENABLE_PREPROCESSING   = YES
-MACRO_EXPANSION        = NO
 EXPAND_ONLY_PREDEF     = NO
 SKIP_FUNCTION_MACROS   = YES
 ALLEXTERNALS           = NO
```